### PR TITLE
Session prepare

### PR DESCRIPTION
**Pull request by nikokaoja**
- added `prepare` endpoint
- added methods to push data model and instances to CDF
- Update docs to say that In Model is manadatory
- Fixed regex patterns 
- Addes specific types for data model external id and space (compliance with CDF API)

### CODE REVIEW

File: cognite/neat/_session/_to.py [hunk 1]
```diff
@@ -35,5 +37,30 @@ def yaml(self, io: Any | None = None) -> str | None:
...
+    def __init__(self, state: SessionState, client: CogniteClient | None, verbose: bool) -> None:
         self._client = client
+        self._state = state
+        self._verbose = verbose
+
+    def instances(self, space: str | None = None):
+        if not self._state.verifies_dms_rules:
+            raise ValueError("No verified DMS data model available")
+
+        space = space or self._state.verifies_dms_rules.metadata.space
```
Comment (doctrino) [inline cognite/neat/_session/_to.py:49]: I think we should not do this. Instead lets prompt the user for instance space if they do not set it.

The `metadata.space` is a schema space and should not be used for data.

File: cognite/neat/_session/_to.py [hunk 2]
```diff
@@ -35,5 +37,30 @@ def yaml(self, io: Any | None = None) -> str | None:
...
+        loader = loaders.DMSLoader.from_rules(self._state.verifies_dms_rules, self._state.store, space)
+
+        if self._client:
+            return loader.load_into_cdf(self._client)
+        else:
+            return loader.write_to_file(Path("./instance.yaml"))
```
Comment (doctrino) [inline cognite/neat/_session/_to.py:55]: Isn't this confusing?  Shouldn't we instead have that placed somewhere else
```
neat.to.file.yaml(format="dms")
```

It is a bit unintuitive to me that if 
```
neat.to.cdf...()
```
I expect to push to CDF.

File: cognite/neat/_session/_to.py [hunk 3]
```diff
@@ -35,5 +37,30 @@ def yaml(self, io: Any | None = None) -> str | None:
...
+
+    def data_model(self):
+        if not self._state.verifies_dms_rules:
+            raise ValueError("No verified DMS data model available")
+
+        exporter = DMSExporter()
+
+        if not self._client:
+            return exporter.export_to_file(self._state.verifies_dms_rules, Path("./data_model.zip"))
```
Comment (doctrino) [inline cognite/neat/_session/_to.py:64]: Same here

File: cognite/neat/_session/_state.py [hunk 4]
```diff
@@ -27,12 +30,36 @@ def input_rule(self) -> ReadRules:
...
+    @property
+    def verifies_dms_rules(self) -> DMSRules | None:
+        if self.verified_rules:
+            for rules in self.verified_rules[::-1]:
+                if isinstance(rules, DMSRules):
+                    return rules
+        return None
+
+    @property
+    def verifies_information_rules(self) -> InformationRules | None:
```
Comment (doctrino) [inline cognite/neat/_session/_state.py:56]: Just to make it clear for us as developers.
```suggestion
    def last_verifies_information_rules(self) -> InformationRules | None:
```

File: cognite/neat/_session/_state.py [hunk 5]
```diff
@@ -27,12 +30,36 @@ def input_rule(self) -> ReadRules:
...
+        return None
+
     @property
     def verified_rule(self) -> VerifiedRules:
         if not self.verified_rules:
             raise ValueError("No verified rules provided")
         return self.verified_rules[-1]
 
+    @property
+    def verifies_dms_rules(self) -> DMSRules | None:
```
Comment (doctrino) [inline cognite/neat/_session/_state.py:48]:
```suggestion
    def last_verifies_dms_rules(self) -> DMSRules | None:
```

File: cognite/neat/_session/_state.py [hunk 6]
```diff
@@ -27,12 +30,36 @@ def input_rule(self) -> ReadRules:
             raise ValueError("No input rules provided")
         return self.input_rules[-1]
 
+    @property
+    def information_input_rule(self) -> ReadRules | None:
```
Comment (doctrino) [inline cognite/neat/_session/_state.py:34]:
```suggestion
    def last_information_input_rule(self) -> ReadRules | None:
```

**Review by doctrino — COMMENTED** [review]
Good stuff. I think it is surprising that you suddenly write to file if you start with `neat.cdf.to`. Wdyt?

File: cognite/neat/_session/_to.py (hunk 1 above)
Comment (nikokaoja) [reply cognite/neat/_session/_to.py:49]: Agree !

File: cognite/neat/_session/_to.py (hunk 2 above)
Comment (nikokaoja) [reply cognite/neat/_session/_to.py:55]: Agree

**Review by doctrino — APPROVED** [review]

**Pull request merged**